### PR TITLE
Fix no-children warnings in objwriter output

### PR DESCRIPTION
**Pull request by agocke**
llvm-dwarfdump produces warnings whenever a debugging entry which is specified to have children has no children. The DWARF spec states that any entry specified to contain children should have at least one child, not including the null entry ending the list of children. It also states that a null entry is allowed, in the place of a child entry.

This change adds a null entry in the case that there are no children for the entry.

### CODE REVIEW

**Comment by agocke** [comment]
@dotnet/illink-contrib for review

**Comment by filipnavara** [comment]
The nicer way to fix it would likely be to:
- Add new abbreviation to `dwarfAbbrev.[cpp/h]` akin to the other "NoChildren" ones
- Update `DwarfClassTypeInfo::DumpTypeInfo` to emit it
- Don't emit the empty list

**Comment by agocke** [comment]
Disagree. I don't see why having an extra abbreviation is better.

**Comment by agocke** [comment]
I just realized that the real answer is probably that we should never be emitting these entries because they're probably useless...

**Comment by filipnavara** [comment]
> Disagree. I don't see why having an extra abbreviation is better.

It would be consistent with how the other entries are emitted. For most programs the output would likely be a bit smaller although I assume the difference in this case would be negligible.

I haven't seen any other compilers emit the null entries, so I am a bit worried that it may cause compatibility issues.

> I just realized that the real answer is probably that we should never be emitting these entries because they're probably useless...

Useless in what sense? And for what debugger / tool? Is the focus strictly lldb? 

At least the fields are used by lldb to print structs/classes. gdb can print types with the `ptype` command and I believe it uses these entries. I also use [LLVM DIVA](https://github.com/SNSystems/DIVA) from time to time.

**Comment by filipnavara** [comment]
Btw, `lldb` can use the information as well:

```
lldb /Users/filipnavara/Projects/runtime/artifacts/tests/coreclr/osx.arm64.Debug/nativeaot/SmokeTests/PInvoke/PInvoke/native/PInvoke          
Cache: /Users/filipnavara/.dotnet/symbolcache
Server: https://msdl.microsoft.com/download/symbols/ Timeout: 4 RetryCount: 0
(lldb) target create "/Users/filipnavara/Projects/runtime/artifacts/tests/coreclr/osx.arm64.Debug/nativeaot/SmokeTests/PInvoke/PInvoke/native/PInvoke"
Current executable set to '/Users/filipnavara/Projects/runtime/artifacts/tests/coreclr/osx.arm64.Debug/nativeaot/SmokeTests/PInvoke/PInvoke/native/PInvoke' (arm64).
(lldb) image lookup --type S_P_CoreLib_System_Environment
0 match found in /Users/filipnavara/Projects/runtime/artifacts/tests/coreclr/osx.arm64.Debug/nativeaot/SmokeTests/PInvoke/PInvoke/native/PInvoke:
id = {0x10005fd72}, name = "S_P_CoreLib_System_Environment", byte-size = 8, compiler_type = "class S_P_CoreLib_System_Environment : private Object {
    static int s_latchedExitCode;
    static int <ProcessorCount>k__BackingField;
    static char s_privilegedProcess;
    static int s_processId;
    static int s_systemPageSize;
    static __Array<> &s_commandLineArgs;
    static String &s_processPath;
    static S_P_CoreLib_System_OperatingSystem &s_osVersion;
    static S_P_CoreLib_System_Collections_Generic_Dictionary_2<> &s_environment;
    static int get_CurrentManagedThreadId();
    static int get_ExitCode();
    static void set_ExitCode(int);
    static void FailFast(String &);
    static void FailFast(String &, S_P_CoreLib_System_Exception &);
    static void FailFast(String &, S_P_CoreLib_System_Exception &, String &);
    static int GetProcessorCount();
    static void ShutdownCore();
    static int get_TickCount();
    static void SetCommandLineArgs(__Array<> &);
    static long get_TickCount64();
    static int get_ProcessorCount();
    static bool get_IsSingleProcessor();
    static String &GetEnvironmentVariable(String &);
    static String &get_NewLine();
    static String &get_StackTrace();
    static String &GetEnvironmentVariableCore(String &);
    static String &TrimStringOnFirstZero(String &);
    static void .cctor();
}"
```

**Comment by agocke** [comment]
> Useless in what sense? And for what debugger / tool? Is the focus strictly lldb?

I think in practical terms should probably never encounter a line of code associated with a completely empty type. So I don't think it should be useful except in the case you're just querying the debug info for some reason.

> It would be consistent with how the other entries are emitted. For most programs the output would likely be a bit smaller although I assume the difference in this case would be negligible.

This is a small benefit and I think it would complicate this code to have to figure out ahead of time if the entry is empty and needs to use the no-children abbrev.

> I haven't seen any other compilers emit the null entries, so I am a bit worried that it may cause compatibility issues.

I actually can't find any examples of other compilers using NoChildren entries either.

**Comment by filipnavara** [comment]
> I think in practical terms should probably never encounter a line of code associated with a completely empty type. So I don't think it should be useful except in the case you're just querying the debug info for some reason.

Agreed. In the case of an empty type it may be fine to omit it entirely. The file/line references for types are not emitted correctly anyway, so the only information it records is whether the type was compiled or not. That information has very marginal usability.

> This is a small benefit and I think it would complicate this code to have to figure out ahead of time if the entry is empty and needs to use the no-children abbrev.

I agree it's harder to read, especially for people not familiar with the DWARF format. It's, however, inconsistent with the rest of the code handling exactly the same case. I don't think it makes the code much more complicated since you basically check for three non-empty lists on two places...

> I actually can't find any examples of other compilers using NoChildren entries either.

The abbreviation table is essentially a dictionary compression. I don't know too many compilers using an explicit static table. In the libraries I used the abbreviation table is generated implicitly by the DWARF library.

**Comment by agocke** [comment]
You're right, it's probably not much more complicated, but I would need to pass down the info to the DumpTypeInfo method since the abbrev entry is not created in the parent method.

Overall, I think this implementation is fine for now, but ideally I'd like to change this code to never emit empty type emtries.

> The abbreviation table is essentially a dictionary compression. I don't know too many compilers using an explicit static table. In the libraries I used the abbreviation table is generated implicitly by the DWARF library.

This is pretty interesting. Do you have a link? I was looking at Go and it looks like they build this table manually.

**Comment by filipnavara** [comment]
> This is pretty interesting. Do you have a link? I was looking at Go and it looks like they build this table manually.

Most experience I have is with https://github.com/xoofx/LibObjectFile/tree/master/src/LibObjectFile/Dwarf which builds the DWARF tree in memory and then generates the abbreviations for you. I think LLVM itself has the same concept with [DIEAbbrevSet](https://github.com/dotnet/llvm-project/blob/dotnet/main/llvm/lib/CodeGen/AsmPrinter/DIE.cpp#L141) and [DIE::computeOffsetsAndAbbrevs](https://github.com/dotnet/llvm-project/blob/5c6667c4823c6609834d0ffb3e88eca2ccf9b526/llvm/lib/CodeGen/AsmPrinter/DIE.cpp#L277-L312) although I am not sure where exactly is it used (ObjWriter doesn't use it at all).

File: llvm/tools/objwriter/debugInfo/dwarf/dwarfTypeBuilder.cpp [hunk 1]
```diff
@@ -399,18 +399,28 @@ void DwarfClassTypeInfo::Dump(UserDefinedDwarfTypesBuilder *TypeBuilder, MCObjec
...
     StaticField.Dump(TypeBuilder, Streamer, TypeSection, StrSection);
   }
 
   for (auto *Function : MemberFunctions) {
+    isEmpty = false;
     Function->Dump(TypeBuilder, Streamer, TypeSection, StrSection);
   }
 
+  if (isEmpty) {
+      // If there are no child nodes we need to write an empty record
```
Comment (tlakollo) [inline llvm/tools/objwriter/debugInfo/dwarf/dwarfTypeBuilder.cpp]: nit: indentation of the comment is off

**Review by tlakollo — APPROVED** [review]

**Review by jtschuster — APPROVED** [review]
Not familiar with DWARF but LGTM

**Comment by agocke** [comment]
This actually didn't work the way I want it to. The DWARF info is significantly different than it was before. Reverting until I can figure out what went wrong.

**Pull request merged**